src/D2L.Bmx/UpdateHandler.cs

```diff
@@ -0,0 +1,130 @@
+using System.Formats.Tar;
+using System.IO.Compression;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace D2L.Bmx;
+
+internal class UpdateHandler {
+
+	public async Task HandleAsync() {
+		using var httpClient = new HttpClient();
+		var releaseData = await GithubUtilities.GetLatestReleaseDataAsync();
+		var localVersion = Assembly.GetExecutingAssembly().GetName().Version;
+		var latestVersion = new Version( GithubUtilities.GetReleaseVersion( releaseData ) );
+		if( latestVersion <= localVersion ) {
+			Console.WriteLine( $"Already own the latest version {latestVersion}" );
+			return;
+		}
+		string archiveName = GetOSFileName();
+		string downloadUrl = releaseData?.Assets?.FirstOrDefault( a => a.Name == archiveName )?.BrowserDownloadUrl
+			?? string.Empty;
+		if( string.IsNullOrWhiteSpace( downloadUrl ) ) {
+			throw new BmxException( "Failed to get update download url" );
+		}
+
+		string downloadPath = Path.GetTempFileName();
+
+		string? currentFilePath = Environment.ProcessPath;
+		if( string.IsNullOrEmpty( currentFilePath ) ) {
+			throw new BmxException( "BMX could not update" );
+		}
+		string currentDirectory = Path.GetDirectoryName( currentFilePath )!;
+		string backupPath = $"{BmxPaths.OLD_BMX_VERSION_FILE_NAME}v{localVersion}.old.bak";
+
+		try {
+			File.Move( currentFilePath, backupPath );
+		} catch( IOException ex ) {
+			throw new BmxException( "Could move the old version, try with elevated permissions", ex );
+		} catch {
+			throw new BmxException( "BMX could not update" );
+		}
+
+		try {
+			var archiveRes = await httpClient.GetAsync( downloadUrl, HttpCompletionOption.ResponseHeadersRead );
+			using( var fs = new FileStream( downloadPath, FileMode.Create, FileAccess.Write, FileShare.None ) ) {
+				await archiveRes.Content.CopyToAsync( fs );
+				await fs.FlushAsync();
+				fs.Dispose();
+			}
+		} catch( Exception ex ) {
+			File.Move( backupPath, currentFilePath );
+			throw new BmxException( "Failed to download the update", ex );
+		}
+
+		try {
+			string extension = Path.GetExtension( downloadUrl ).ToLowerInvariant();
+
+			if( extension.Equals( ".zip" ) ) {
```

Comment:
instead of calling `ToLowerInvariant`, use [Equals(String, StringComparison)](https://learn.microsoft.com/en-us/dotnet/api/system.string.equals?view=net-8.0#system-string-equals(system-string-system-stringcomparison)) with `StringComparison.OrdinalIgnoreCase`